examples/nlp/text_generation_gpt.py

```diff
@@ -0,0 +1,418 @@
+are all not useful for this text corpus and unneccesarily increases vocabulary size. Then
+we take the `VOCAB_SIZE` most frequent vocabularies after that. There are also some
+reserved tokens, so we want to make sure that `"[PAD]"` is at index 0 and `"[UNK]"` is in
+the vocabulary, since it is used for padding and unknown words respectively.
+"""
+
+# Download vocabulary data.
+vocab_file = keras.utils.get_file(
+    origin="https://storage.googleapis.com/tensorflow/keras-nlp/examples/bert/bert_vocab_uncased.txt",
+)
+# Get most frequent VOCAB_SIZE vocab tokens for faster/efficient training.
+vocab_list = []
+with open(vocab_file, "r") as f:
+    for line in f:
+        vocab_list.append(line.strip())
+
+# The first 1000 tokens are unused tokens that unneccesarily increases vocab size.
+vocab_list = ["[PAD]", "[UNK]", "[CLS]", "[SEP]", "[MASK]"] + vocab_list[1000:]
+vocab_list = vocab_list[:VOCAB_SIZE]
+
+tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=vocab_list,
+    sequence_length=SEQ_LEN + 1,  # +1 for slicing feature and label.
+)
+
+"""
+## Load Data
+
+Now, let's download the dataset! The SimpleBooks dataset has a small vocabulary, which
+makes it easier to fit a simple model on.
+"""
+
+keras.utils.get_file(
+    origin="https://dldata-public.s3.us-east-2.amazonaws.com/simplebooks.zip",
+    extract=True,
+)
+dir = os.path.expanduser("~/.keras/datasets/simplebooks/")
+
+# Load simplebooks-92 train set and filter out short lines.
+train_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/train.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+    .shuffle(buffer_size=256)
+)
+
+# Load simplebooks-92 validation set and filter out short lines.
+valid_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/valid.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+)
+
+"""
+## Tokenize Data
+
+We pre-process the dataset by tokenizing and splitting it into `features` and `labels`.
+"""
+
+
+def preprocess(inputs):
+    outputs = tokenizer(inputs)
+    features = outputs[:, :-1]
+    labels = outputs[:, 1:]
+    return features, labels
+
+
+# Tokenize and split into train and label sequences.
+ds = train_ds.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
+    tf.data.AUTOTUNE
+)
+ds_valid = valid_ds.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
+    tf.data.AUTOTUNE
+)
+
+"""
+## Model
+
+We create our scaled down GPT model with the following layers:
+
+- One `keras_nlp.layers.TokenAndPositionEmbedding` layer, which combines the embedding
+for the token and its position. 
+- Multiple `keras_nlp.layers.TransformerDecoder` layers, with the default causal masking.
+The layer has no cross-attention when run with decoder sequence only.
+- One output dense linear layer
+"""
+
+
+def create_model():
+    inputs = keras.layers.Input(shape=(SEQ_LEN,), dtype=tf.int32)
```

Comment:
given that this is only called once, why not just move this out of the function?

Reply:
edited